compiler/rustc_codegen_ssa/src/base.rs

```diff
@@ -380,7 +380,19 @@ pub fn memcpy_ty<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
         return;
     }
 
-    bx.memcpy(dst, dst_align, src, src_align, bx.cx().const_usize(size), flags);
+    if flags == MemFlags::empty()
+        && let Some(bty) = bx.cx().scalar_copy_backend_type(layout)
+    {
+        // I look forward to only supporting opaque pointers
+        let pty = bx.type_ptr_to(bty);
+        let src = bx.pointercast(src, pty);
+        let dst = bx.pointercast(dst, pty);
```

Comment:
When do these not match pty? Why not just return a more "accurate" type in `scalar_copy_llvm_type`? Like the actual LLVM type that corresponds to src/dst? (Or am I misunderstanding something?)

Reply:
(Well, the shallow answer is that they don't match because it's `-> Option<Type>`, not `should_load_store_instead_of_memcpy() -> bool`.)

Because emitting these as the backend type doesn't necessarily do what we want.  Notably, this PR is emitting the load/store as `<4 x i8>`, the *vector* type, rather than the `llvm_backend_type` of `[4 x i8]` for arrays.

I tried using the LLVM type first, but with arrays that results in exploding out the IR: <https://llvm.godbolt.org/z/vjjsdea9e>

Optimizing the version that loads/stores arrays
```llvm
define void @replace_short_array_using_arrays(ptr noalias nocapture noundef sret([3 x i32]) dereferenceable(12) %0, ptr noalias noundef align 4 dereferenceable(12) %r, ptr noalias nocapture noundef readonly dereferenceable(12) %v) unnamed_addr #0 {
start:
  %1 = load [3 x i32], ptr %r, align 4
  store [3 x i32] %1, ptr %0, align 4
  %2 = load [3 x i32], ptr %v, align 4
  store [3 x i32] %2, ptr %r, align 4
  ret void
}
```
gives
```llvm
define void @replace_short_array_using_arrays(ptr noalias nocapture noundef writeonly sret([3 x i32]) dereferenceable(12) %0, ptr noalias nocapture noundef align 4 dereferenceable(12) %r, ptr noalias nocapture noundef readonly dereferenceable(12) %v) unnamed_addr #0 {
  %.unpack = load i32, ptr %r, align 4
  %.elt1 = getelementptr inbounds [3 x i32], ptr %r, i64 0, i64 1
  %.unpack2 = load i32, ptr %.elt1, align 4
  %.elt3 = getelementptr inbounds [3 x i32], ptr %r, i64 0, i64 2
  %.unpack4 = load i32, ptr %.elt3, align 4
  store i32 %.unpack, ptr %0, align 4
  %.repack5 = getelementptr inbounds [3 x i32], ptr %0, i64 0, i64 1
  store i32 %.unpack2, ptr %.repack5, align 4
  %.repack7 = getelementptr inbounds [3 x i32], ptr %0, i64 0, i64 2
  store i32 %.unpack4, ptr %.repack7, align 4
  %.unpack9 = load i32, ptr %v, align 4
  %.elt10 = getelementptr inbounds [3 x i32], ptr %v, i64 0, i64 1
  %.unpack11 = load i32, ptr %.elt10, align 4
  %.elt12 = getelementptr inbounds [3 x i32], ptr %v, i64 0, i64 2
  %.unpack13 = load i32, ptr %.elt12, align 4
  store i32 %.unpack9, ptr %r, align 4
  store i32 %.unpack11, ptr %.elt1, align 4
  store i32 %.unpack13, ptr %.elt3, align 4
  ret void
}
```
whereas optimizing the version with vectors leaves the operations together
```
define void @replace_short_array_using_vectors(ptr noalias nocapture noundef writeonly sret([3 x i32]) dereferenceable(12) %0, ptr noalias nocapture noundef align 4 dereferenceable(12) %r, ptr noalias nocapture noundef readonly dereferenceable(12) %v) unnamed_addr #0 {
  %1 = load <3 x i32>, ptr %r, align 4
  store <3 x i32> %1, ptr %0, align 4
  %2 = load <3 x i32>, ptr %v, align 4
  store <3 x i32> %2, ptr %r, align 4
  ret void
}
```
for the backend to legalize instead.

Reply:
Oh, and I originally expected to do this with LLVM's arbitrary-length integer support <https://llvm.godbolt.org/z/hzG9aeqhM>, like I did for comparisons back in #85828
```llvm
define void @replace_short_array_using_long_integer(ptr noalias nocapture noundef sret([3 x i32]) dereferenceable(12) %0, ptr noalias noundef align 4 dereferenceable(12) %r, ptr noalias nocapture noundef readonly dereferenceable(12) %v) unnamed_addr #0 {
start:
  %1 = load i96, ptr %r, align 4
  store i96 %1, ptr %0, align 4
  %2 = load i96, ptr %v, align 4
  store i96 %2, ptr %r, align 4
  ret void
}
```

But that broke some of the autovectorization codegen tests for operations on arrays.

Reply:
Yeah ok, so we explicitly want to be lowering these moves to vector types specifically because they can optimize better than arrays.